File: 143.重排链表.go
/*
 * @lc app=leetcode.cn id=143 lang=golang
 *
 * [143] 重排链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode)  {
	if head == nil || head.Next == nil {
		return
	}

	// 寻找中间节点
	middle, fast := head, head
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		middle = middle.Next
	}

	// 反转
	var pre *ListNode
	node := middle
	for node != nil {
		tmp := node.Next
		node.Next = pre
		pre = node
		node = tmp
	}

	// 合并
	cur1, cur2 := head, pre
	for cur2.Next != nil {
		tmp1 := cur1.Next
		cur1.Next = cur2
		cur1 = tmp1;

		tmp2 := cur2.Next
		cur2.Next = cur1
		cur2 = tmp2
	}
}
// @lc code=end

File: 143.重排链表.ts
/*
 * @lc app=leetcode.cn id=143 lang=typescript
 *
 * [143] 重排链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

/**
 Do not return anything, modify head in-place instead.
 */
function reorderList(head: ListNode | null): void {
    if (!head || !head.next) return;

    let [middle, fast] = [head, head];
    while (fast && fast.next) {
        fast = fast.next.next;
        middle = middle.next;
    }
    
    // 反转
    let pre: ListNode | null = null;
    let node = middle;
    while (node) {
        let tmp = node.next;
        node.next = pre;
        pre = node;
        node = tmp;
    }

    // 合并
    let [cur1, cur2] = [head, pre];
    while (cur2!.next) {
        let first = cur1!.next;
        cur1.next = cur2!;
        cur1 = first;

        let second = cur2!.next;
        cur2!.next = cur1!;
        cur2 = second;
    }
};
// @lc code=end

